lamindb_setup/dev/upath.py

```diff
@@ -18,16 +18,12 @@
 from lamin_utils import logger
 from upath import UPath
 from upath.implementations.cloud import CloudPath, S3Path  # noqa
+from upath.implementations.local import LocalPath, PosixUPath, WindowsUPath
 
-AWS_CREDENTIALS_PRESENT = None
+LocalPathClasses = (PosixUPath, WindowsUPath, LocalPath)
```

Comment:
I think I read up in the `universal_pathlib` code base and saw that the class hierarchy is such that this is possible.

Meaning `PosixPath` and `WindowsPath` inherit from `LocalPath` and `S3Path` and `GCPPath` inherit from `CloudPath`.

Unfortunately, it seems, I didn't paste a link in our thread: https://laminlabs.slack.com/archives/C04FPE8V01W/p1691069418969499?thread_ts=1691059791.005409&cid=C04FPE8V01W

Reply:
Hm, i checked the code and also directly with `isinstance`, they don't inherit. I can recheck.

Reply:
https://github.com/fsspec/universal_pathlib/blob/main/upath/implementations/local.py
The code is here, no inheritance.

Reply:
I am not really sure.
`UPath("xyz.xyz")` gives `PosixUPath` or `WindowsUPath`, .
And `UPath("file://xyz.xyz")` gives `LocalPath`.

We can ask the authors directly maybe.

Reply:
Understood! In that case, I think they misnamed this class. They should rather have chosen a more specific name.

Also, it's clear from the definition universal_pathlib's LocalPath behaves just like UPath (and it that sense, I was wrong in calling it "abstract baseclass"):

```
class LocalPath(UPath):
    pass
```

I don't think we want to support cases like `file://xyz.xyz` for which the only thing that `UPath` seems to be doing is calling the `LocalPath` constructor during initialization.

I think what we want is a simple & predictable framework that makes the code readable and enables users & developers to understand what's going on. And for that, "local path" vs. "cloud path" seems the right level of abstraction.

Please make an issue within UPath! Maybe you're right and they'll continue to use `LocalPath` for species non-posix and non-windows cases also in the future. At that point, we could update our implementation if we need. We could also discuss with them the intentions behind the naming choice. If this is how they want to use `LocalPath`, I think they should consider renaming it and introduce a parent class `LocalPath`.

Until they get back, I think, we should work with a simple set of sensibly named classes and `universal_pathlib.implementations.local.LocalPath` just seems to introduce unnecessary confusion and complexity.

Let's call if you're not convinced! 😄

Reply:
Yes. One additional discussion point here. I made it more general now, so that all filesystems supported by `UPath` are allowed. Should i maybe throw an error if not `UPath` is not `CloudPath` or `LocalPathClasses`? In principle it is not needed, but we might want to restrict them intentionally.

Reply:
🤔 I'd need to think more about what all implications are - I'm leaning towards not throwing an error right now.

Reply:
I don't think not restricting is problematic. We just don't list any filesystems that we don't test as supported. And even restircting to `CloudPath` and `LocalPathClasses` would still allow Azure that we don't test.

Reply:
Sounds good!